InfluxDB: Implement InfluxQL json streaming parser

itsmylife opened the pull request:
**What is this feature?**

The initial attempt to implement a streaming parser for InfluxDB influxql responses.

`Buffered` is the current response parser we have.
`Stream` is the new parser we introduce with this PR.
```
// Big response ~15mb 

// time_series response format
pkg: github.com/grafana/grafana/pkg/tsdb/influxdb/influxql
             │ buffered2.txt │             stream2.txt             │
             │    sec/op     │   sec/op     vs base                │
ParseJson-10     454.3m ± 2%   307.6m ± 2%  -32.30% (p=0.000 n=10)

             │ buffered2.txt │             stream2.txt              │
             │     B/op      │     B/op      vs base                │
ParseJson-10    488.4Mi ± 0%   382.7Mi ± 0%  -21.66% (p=0.000 n=10)

             │ buffered2.txt │             stream2.txt             │
             │   allocs/op   │  allocs/op   vs base                │
ParseJson-10     9.588M ± 0%   6.689M ± 0%  -30.24% (p=0.000 n=10)


// table response format
pkg: github.com/grafana/grafana/pkg/tsdb/influxdb/influxql
             │ bufferedt.txt │             streamt.txt             │
             │    sec/op     │   sec/op     vs base                │
ParseJson-10    4504.1m ± 1%   176.5m ± 1%  -96.08% (p=0.000 n=10)

             │ bufferedt.txt  │             streamt.txt              │
             │      B/op      │     B/op      vs base                │
ParseJson-10   31689.6Mi ± 0%   277.1Mi ± 0%  -99.13% (p=0.000 n=10)

             │ bufferedt.txt │             streamt.txt             │
             │   allocs/op   │  allocs/op   vs base                │
ParseJson-10     3.406M ± 0%   4.029M ± 0%  +18.28% (p=0.000 n=10)
```

**Why do we need this feature?**

For better response parsing

**Who is this feature for?**

InfluxDB InfluxQL users

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/65044

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

itsmylife committed:
Have the first iteration

itsmylife committed:
Prepare bench testing

ryantxu commented on pkg/tsdb/influxdb/influxql/converter/testdata/influx_select_*_from_cpu-frame.jsonc:
can we keep "*" out of the file name?  It may work sometimes, but i expect it to cause problems somewhere at some point and not be something we are hoping to debug :)

itsmylife replied:
Definitely!

ryantxu commented on pkg/tsdb/influxdb/influxql/converter/testdata/influx_select_value_from_*-frame.jsonc:
```diff
@@ -0,0 +1,128 @@
+//  Name: cpu.mean
+//  Dimensions: 2 Fields by 4 Rows
+//  +-------------------------------+-------------------+
+//  | Name: Time                    | Name: Value       |
```
when looking at the input vs output, i would expect the parser to produce a frame with name "cpu" and fields matching the names in:
```
          "name": "cpu",
          "columns": [
            "time",
            "mean_usage_guest",
            "mean_usage_nice",
            "mean_usage_idle"
          ],
```
if we then need to split that in to [timeseries multi](https://grafana.github.io/dataplane/timeseries#time-series-multi-format-timeseriesmulti), that is a straightforward task (but should not be done as part of parsing)

ryantxu commented on pkg/tsdb/influxdb/influxql/converter/testdata/influx_select_*_from_cpu-frame.jsonc:
```diff
@@ -0,0 +1,166 @@
+//  🌟 This was machine generated.  Do not edit. 🌟
+//  
+//  Frame[0] 
+//  Name: cpu.mean_usage_guest
```
when looking at the input vs output, i would expect the parser to produce a frame with name "cpu" and fields matching the names in:
```
          "name": "cpu",
          "columns": [
            "time",
            "mean_usage_guest",
            "mean_usage_nice",
            "mean_usage_idle"
          ],
```
if we then need to split that in to [timeseries multi](https://grafana.github.io/dataplane/timeseries#time-series-multi-format-timeseriesmulti), that is a straightforward task (but should not be done as part of parsing)

itsmylife replied:
I can produce the result that way. Is that what we want to achieve in general?

ryantxu replied:
In general i think there should be the fewest *changes* to from the input format to the output.  The input values here are structured as a table, so keep them as a table :)

itsmylife replied:
This is not a table. This is a general response from InfluxDB. See here https://docs.influxdata.com/influxdb/v2/api/#operation/GetLegacyQuery 
To able to create timeseries data we were converting it to data frames already. 

On Frontend Mode
- https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/influxdb/datasource.ts#L629-L634
- https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/influxdb/influx_series.ts#L21-L67

On Backend Mode
- https://github.com/grafana/grafana/blob/main/pkg/tsdb/influxdb/influxql/response_parser.go#L77-L115

ryantxu commented on pkg/tsdb/influxdb/influxql/converter/testdata/influx_select_value_from_*-frame.jsonc:
```diff
@@ -0,0 +1,128 @@
+//  Frame[1] 
+//  Name: derivative.mean
+//  Dimensions: 0 Fields by 0 Rows
+//  +
```
this looks like it is dropping the additional results

itsmylife replied:
I don't get that. Which one is it dropping?

itsmylife replied:
I fixed this in latest commits.

itsmylife committed:
Merge branch 'main' into ismail/influxdb-json-streaming-parser

itsmylife committed:
rename the test files

itsmylife committed:
Remove unnecessary test file

itsmylife committed:
Introduce influxqlStreamingParser feature flag

itsmylife committed:
Apply streaming parser feature flag

itsmylife committed:
Add new tests

itsmylife committed:
More tests

itsmylife committed:
return executedQueryString only in first frame

itsmylife committed:
add frame meta and config

itsmylife committed:
Update golden json files

itsmylife commented on pkg/tsdb/influxdb/influxql/response_parser.go:
```diff
@@ -204,11 +204,13 @@ func newFrameWithoutTimeField(row models.Row, retentionPolicyQuery bool, tagValu
-	frame.Meta = &data.FrameMeta{
-		ExecutedQueryString:    queryString,
-		PreferredVisualization: visType,
+	if frameIndex == 0 {
+		frame.Meta = &data.FrameMeta{
+			ExecutedQueryString:    queryString,
+			PreferredVisualization: visType,
+		}
```
We did the same thing for Prometheus to send less data to the frontend.

itsmylife commented on pkg/tsdb/influxdb/influxql/converter/converter_test.go:
```diff
@@ -1,47 +1,100 @@
+
+		if streamParser {
+			// shouldUpdate = false
+			rsp = influxql.StreamParse(f, 200, query)
+		} else {
+			rsp = *influxql.ResponseParse(io.NopCloser(f), 200, query)
```
So current response parser returns a pointer and the new parser doesn't. I am not sure which one is better. (A bit confused too). I am open to suggestions.

itsmylife committed:
Support tags/labels

itsmylife committed:
more tests

itsmylife committed:
more tests

itsmylife committed:
Merge branch 'main' into ismail/influxdb-json-streaming-parser

# Conflicts:
#	pkg/tsdb/influxdb/influxql/response_parser.go

itsmylife committed:
Merge branch 'main' into ismail/influxdb-json-streaming-parser

# Conflicts:
#	docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md
#	packages/grafana-data/src/types/featureToggles.gen.ts
#	pkg/services/featuremgmt/registry.go
#	pkg/services/featuremgmt/toggles_gen.csv
#	pkg/services/featuremgmt/toggles_gen.go
#	pkg/tsdb/influxdb/influxql/response_parser.go

itsmylife committed:
Don't change original response_parser.go

itsmylife committed:
provide context

itsmylife committed:
create util package

itsmylife committed:
don't pass the row

itsmylife committed:
update converter with formatted frameName

itsmylife committed:
add executedQueryString info only to first frame

itsmylife committed:
update golden files

itsmylife committed:
rename

itsmylife committed:
update test file

itsmylife committed:
use pointer values

itsmylife committed:
update testdata

itsmylife committed:
update parsing

itsmylife committed:
update converter for null values

itsmylife committed:
prepare converter for table response

itsmylife committed:
clean up

itsmylife committed:
return timeField in fields

itsmylife committed:
handle no time column responses

itsmylife committed:
better nil field handling

itsmylife committed:
refactor the code

kylebrandt commented on pkg/tsdb/influxdb/influxql/converter/converter.go:
```diff
@@ -0,0 +1,329 @@
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	jsoniter "github.com/json-iterator/go"
```
I think we should wrap whatever we need from this import into `jsonitere` so this import is used. That way there won't be a mistake of calling jsoniter (and skipping err lint) when working on the code.

kylebrandt replied:
Still think this is worth doing

itsmylife replied:
I use this to create the initial `jsonitere` instance on line 22
```
iter := jsonitere.NewIterator(jIter)
```
So I think it is ok to stay like this. WDYT?

itsmylife committed:
add table tests

itsmylife committed:
fix config for table

itsmylife committed:
table response format

itsmylife committed:
fix value

itsmylife committed:
if there is no time column set name

itsmylife committed:
linting

itsmylife committed:
refactoring

itsmylife committed:
handle the status code

itsmylife committed:
add tracing

inancgumus commented on pkg/tsdb/influxdb/influxql/converter/converter.go:
```diff
@@ -0,0 +1,395 @@
+}
+
+func readSeries(iter *jsonitere.Iterator, frameName []byte, query *models.Query) *backend.DataResponse {
+	var measurement string
+	var tags map[string]string
+	var columns []string
+	var valueFields data.Fields
+	var hasTimeColumn bool
```
Just a style suggestion for better readability:

```suggestion
	var (
		measurement   string
		tags          map[string]string
		columns       []string
		valueFields   data.Fields
		hasTimeColumn bool
	)
```

inancgumus commented on pkg/tsdb/influxdb/influxql/converter/converter.go:
```diff
@@ -0,0 +1,395 @@
+				if err != nil {
+					return rspErr(err)
+				}
+				fmt.Printf("unsupported key: %s / %v\n", l1Field, v)
```
Is using a `Printf` OK?

inancgumus replied:
Since `Printf` writes directly to `stdout`, in other projects, for example, we don't use it. So that we can control the output from tests and even change it.

itsmylife replied:
it is ok for now. I'll keep it.

inancgumus commented on pkg/tsdb/influxdb/influxql/converter/converter.go:
```diff
@@ -0,0 +1,395 @@
+			handleTableFormatFirstFrame(rsp, measurement, query)
+			handleTableFormatFirstField(rsp, valueFields, columns)
+			handleTableFormatTagFields(rsp, valueFields, tags)
+			handleTableFormatValueFields(rsp, valueFields, tags, columns)
```
I would return `rsp` here, and remove the `else` block below. Benefits: better readability and reduced cognitive overload caused by indentation.

itsmylife replied:
I cannot return the rsp here. It is still in the loop.

inancgumus commented on pkg/tsdb/influxdb/influxql/converter/converter.go:
```diff
@@ -0,0 +1,395 @@
+			if hasTimeColumn {
+				// Frame with time column
+				newFrames := handleTimeSeriesFormatWithTimeColumn(valueFields, tags, columns, measurement, frameName, query)
+				rsp.Frames = append(rsp.Frames, newFrames...)
```
I would return `rsp` also here and remove the `else` below.

itsmylife replied:
Cannot return rsp here either

inancgumus commented on pkg/tsdb/influxdb/influxql/converter/converter.go:
```diff
@@ -0,0 +1,395 @@
+}
+
+func readColumns(iter *jsonitere.Iterator) (columns []string, err error) {
+	columns = make([]string, 0)
```
Is this necessary? `columns` is already appendable. I would use an explicit return instead of a naked one for clarity: `return columns`.

inancgumus commented on pkg/tsdb/influxdb/influxql/converter/converter.go:
```diff
@@ -0,0 +1,395 @@
+}
+
+func readValues(iter *jsonitere.Iterator, hasTimeColumn bool) (valueFields data.Fields, err error) {
+	valueFields = make(data.Fields, 0)
```
Similar to my previous suggestion, is initializing `valueFields` necessary?

inancgumus commented on pkg/tsdb/influxdb/influxql/converter/converter.go:
```diff
@@ -0,0 +1,395 @@
+				if t, err = iter.ReadFloat64(); err != nil {
+					return nil, err
+				}
+				valueFields[0].Append(time.UnixMilli(int64(t)).UTC())
```
I would use `colIdx++; continue` here and remove the following `else`.

inancgumus commented on pkg/tsdb/influxdb/influxql/converter/converter.go:
```diff
@@ -0,0 +1,395 @@
+// For nil values we might have added NullableFloat64 value field
+// if they are not matching fix it
+func maybeFixValueFieldType(valueFields data.Fields, expectedType data.FieldType, colIdx int) {
+	if valueFields[colIdx].Type() != expectedType {
+		stringField := data.NewFieldFromFieldType(expectedType, 0)
+		stringField.Name = "Value"
+		var i int
+		for i < valueFields[colIdx].Len() {
+			stringField.Append(nil)
+			i++
+		}
+		valueFields[colIdx] = stringField
```
```suggestion
	if valueFields[colIdx].Type() == expectedType {
		return
	}
	stringField := data.NewFieldFromFieldType(expectedType, 0)
	stringField.Name = "Value"
	for i := 0; i < valueFields[colIdx].Len(); i++ {
		stringField.Append(nil)
	}
	valueFields[colIdx] = stringField
```

inancgumus commented on pkg/tsdb/influxdb/influxql/converter/converter.go:
```diff
@@ -0,0 +1,395 @@
+func handleTimeSeriesFormatWithoutTimeColumn(valueFields data.Fields, columns []string, measurement string, query *models.Query) *data.Frame {
+	// Frame without time column
+	var frame *data.Frame
+	if strings.Contains(strings.ToLower(query.RawQuery), strings.ToLower("SHOW TAG VALUES")) {
+		if len(columns) >= 2 {
+			frame = data.NewFrame(measurement, valueFields[1])
+		}
+	} else {
+		if len(columns) >= 1 {
+			frame = data.NewFrame(measurement, valueFields[0])
+		}
+	}
+	return frame
```
```suggestion
	if len(columns) >= 2 && strings.Contains(strings.ToLower(query.RawQuery), strings.ToLower("SHOW TAG VALUES")) {
		return data.NewFrame(measurement, valueFields[1])
	}
	if len(columns) >= 1 {
		return data.NewFrame(measurement, valueFields[0])
	}
	return nil
```

inancgumus commented on pkg/tsdb/influxdb/influxql/converter/converter_test.go:
```diff
@@ -0,0 +1,92 @@
+//lint:ignore U1000 Ignore used function for now
+func runScenario(tf string, resultFormat string) func(t *testing.T) {
+	return func(t *testing.T) {
+		// Safe to disable, this is a test.
+		// nolint:gosec
+		f, err := os.Open(path.Join("testdata", tf+".json"))
```
```suggestion
		f, err := os.Open(path.Join("testdata", filepath.Clean(tf+".json")))
```

inancgumus commented on pkg/tsdb/influxdb/influxql/util/util.go:
```diff
@@ -0,0 +1,154 @@
+	nameSegment := strings.Split(rowName, ".")
+
+	result := legendFormat.ReplaceAllFunc([]byte(query.Alias), func(in []byte) []byte {
+		aliasFormat := string(in)
+		aliasFormat = strings.Replace(aliasFormat, "[[", "", 1)
+		aliasFormat = strings.Replace(aliasFormat, "]]", "", 1)
+		aliasFormat = strings.Replace(aliasFormat, "$", "", 1)
```
If performance is a concern, I would move over the bytes and replace them in a loop. This creates a lot of strings. I haven't measured it, though :)

itsmylife replied:
I decided to create a new PR that introduces the util package. Those functions were in the code for a while already. I moved them into this package. For the performance implications, I'll have another PR.

inancgumus commented on pkg/tsdb/influxdb/influxql/util/util.go:
```diff
@@ -0,0 +1,154 @@
+	}
+	frameName = append(frameName, column...)
+
+	if len(tags) > 0 {
+		frameName = append(frameName, ' ', '{', ' ')
+		first := true
+		for k, v := range tags {
+			if !first {
+				frameName = append(frameName, ',')
+				frameName = append(frameName, ' ')
+			} else {
+				first = false
+			}
+			frameName = append(frameName, k...)
+			frameName = append(frameName, ':', ' ')
+			frameName = append(frameName, v...)
+		}
+
+		frameName = append(frameName, ' ', '}')
+	}
+
+	return frameName
```
```suggestion
	if len(tags) == 0 {
		return frameName
	}
	frameName = append(frameName, ' ', '{', ' ')
	first := true
	for k, v := range tags {
		if !first {
			frameName = append(frameName, ',')
			frameName = append(frameName, ' ')
		} else {
			first = false
		}
		frameName = append(frameName, k...)
		frameName = append(frameName, ':', ' ')
		frameName = append(frameName, v...)
	}
	return append(frameName, ' ', '}')
```

itsmylife replied:
I have this in https://github.com/grafana/grafana/pull/78826

inancgumus reviewed (approved):
Hi! Mostly, LGTM 👏 Remember that I don't know about this project, so take my approval with a grain of salt 😄

I've included some stylistic suggestions. Feel free to apply to the rest of the code. I haven't included similar suggestions for the rest of the code.

itsmylife committed:
Update pkg/tsdb/influxdb/influxql/converter/converter_test.go

Co-authored-by: İnanç Gümüş <m@inanc.io>

itsmylife committed:
Merge branch 'main' into ismail/influxdb-json-streaming-parser

# Conflicts:
#	pkg/tsdb/influxdb/influxql/response_parser.go
#	pkg/tsdb/influxdb/influxql/response_parser_test.go

itsmylife committed:
Merge branch 'main' into ismail/influxdb-json-streaming-parser

# Conflicts:
#	pkg/tsdb/influxdb/influxql/util/util.go

itsmylife committed:
fix import

itsmylife committed:
Merge branch 'main' into ismail/influxdb-json-streaming-parser

# Conflicts:
#	docs/sources/setup-grafana/configure-grafana/feature-toggles/index.md
#	packages/grafana-data/src/types/featureToggles.gen.ts
#	pkg/services/featuremgmt/registry.go
#	pkg/services/featuremgmt/toggles_gen.csv
#	pkg/services/featuremgmt/toggles_gen.go

itsmylife committed:
update test data

itsmylife committed:
sanity

itsmylife committed:
sanity

itsmylife committed:
linting

itsmylife committed:
simplicity

itsmylife committed:
return empty rsp

itsmylife committed:
rename to prevent confusion

itsmylife committed:
nullableJson field type for null values

itsmylife committed:
better handling null values

itsmylife committed:
remove duplicate test file

itsmylife committed:
fix healthcheck

itsmylife committed:
Merge branch 'main' into ismail/influxdb-json-streaming-parser

itsmylife committed:
use util for pointer

itsmylife committed:
move bench test to root

itsmylife committed:
provide fake feature manager

kylebrandt commented on pkg/tsdb/influxdb/influxql/converter/converter.go:
```diff
@@ -0,0 +1,423 @@
+				maybeFixValueFieldType(valueFields, data.FieldTypeNullableBool, colIdx)
+				tryToAppendValue(valueFields, util.ToPtr(b.ToBool()), colIdx)
+			case jsoniter.NilValue:
+				_, _ = iter.Read()
```
Why not checking err or response here? comment maybe?

itsmylife replied:
It is a `null` value already. should we check regardless?

itsmylife commented on pkg/tsdb/influxdb/healthcheck.go:
```diff
@@ -7,8 +7,10 @@ import (
+	"github.com/grafana/grafana-plugin-sdk-go/backend/tracing"
 
 	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
```
I agree with you but I think our tooling is not ready for this, yet
```
wire: /Users/ismailsimsek/Documents/workspace/grafana/pkg/server/wire.go:432:1: inject Initialize: no provider found for github.com/grafana/grafana-plugin-sdk-go/experimental/featuretoggles.FeatureToggles
```

itsmylife committed:
Merge branch 'main' into ismail/influxdb-json-streaming-parser

itsmylife committed:
add more tests

itsmylife committed:
partial fix for null values in table response format

itsmylife committed:
handle partial null fields

itsmylife committed:
comments for easy testing

itsmylife committed:
move frameName allocation in readSeries

itsmylife committed:
one less append operation

itsmylife committed:
performance improvement by making string conversion once

pkg: github.com/grafana/grafana/pkg/tsdb/influxdb/influxql
             │ stream2.txt │            stream3.txt             │
             │   sec/op    │   sec/op     vs base               │
ParseJson-10   314.4m ± 1%   303.9m ± 1%  -3.34% (p=0.000 n=10)

             │ stream2.txt  │             stream3.txt              │
             │     B/op     │     B/op      vs base                │
ParseJson-10   425.2Mi ± 0%   382.7Mi ± 0%  -10.00% (p=0.000 n=10)

             │ stream2.txt │            stream3.txt             │
             │  allocs/op  │  allocs/op   vs base               │
ParseJson-10   7.224M ± 0%   6.689M ± 0%  -7.41% (p=0.000 n=10)

itsmylife committed:
add comment lines
